### PR TITLE
feat(providers/openai): Add full reasoning support to OpenAI responses API

### PR DESCRIPTION
**Pull request by jonaslalin**
## Background

This PR adds comprehensive support for OpenAI's Responses API with full reasoning capabilities. The OpenAI Responses API is a new API that provides access to OpenAI's reasoning models (o1, o3, o3-mini, o4-mini) with enhanced reasoning summaries, encrypted content handling, and zero data retention options.

## Summary

This PR implements full reasoning support for the OpenAI Responses API, including:

- **Async message conversion**: Updated `convertToOpenAIResponsesMessages` to handle asynchronous reasoning message processing with provider options parsing
- **Reasoning message handling**: Added comprehensive support for reasoning parts with ID generation, encrypted content handling, and summary merging
- **Streaming support**: Full streaming implementation for reasoning summaries with encrypted content support
- **Enhanced API types**: Added `OpenAIResponsesReasoning` type with encrypted content and summary structures
- **Provider options**: Added support for `reasoningSummary`, `include`, and reasoning effort configuration
- **Zero data retention**: Support for stateless interactions with encrypted content retrieval
- **Comprehensive examples**: Added 4 new examples demonstrating different reasoning use cases
- **Robust error handling**: Added validation and warnings for malformed reasoning parts and mismatched encrypted content

### Key Changes:
- Made `convertToOpenAIResponsesMessages` async to support provider options parsing
- Added reasoning message merging logic for consecutive parts with the same ID
- Enhanced streaming to include reasoning-specific events and metadata
- Updated usage token handling to include reasoning tokens in the standard usage object
- Added comprehensive test coverage for all reasoning scenarios

### Manual Testing Performed:
1. **Basic reasoning generation**: Tested with o3-mini model generating reasoning summaries
2. **Streaming reasoning**: Verified real-time streaming of reasoning content with proper start/end events
3. **Encrypted content handling**: Tested zero data retention scenarios with encrypted content retrieval
4. **Message conversion**: Verified proper conversion of reasoning parts to OpenAI API format
5. **Error scenarios**: Tested malformed provider options and mismatched encrypted content handling
6. **Multi-step reasoning**: Tested complex reasoning flows with tool interactions
7. **All 4 new examples**: Ran and verified all example scripts work correctly with the API

### Test Coverage:
- Added 15+ new test cases covering reasoning message conversion
- Added streaming tests for reasoning with encrypted content
- Added comprehensive error handling tests
- All existing tests continue to pass

## Tasks

- [x] Tests have been added / updated (for bug fixes / features)
- [x] Documentation has been added / updated (for bug fixes / features) - via examples
- [x] A _patch_ changeset for relevant packages has been added (for bug fixes / features - run `pnpm changeset` in the project root)
- [x] Formatting issues have been fixed (run `pnpm prettier-fix` in the project root)

## Related Issues

Fixes [#6488](https://github.com/vercel/ai/issues/6488)

### CODE REVIEW

**Comment by jonaslalin** [comment]
@lgrammel can you please review. I have seen that you have been working in these files previously :)

File: examples/ai-core/src/generate-text/openai-responses-reasoning-summary-streaming.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
+import { openai, OpenAIResponsesProviderOptions } from '@ai-sdk/openai';
+import { streamText } from 'ai';
+import 'dotenv/config';
+
+async function main() {
```
Comment (lgrammel) [inline examples/ai-core/src/generate-text/openai-responses-reasoning-summary-streaming.ts]: have you tested this with streamText as well?

File: packages/openai/src/responses/openai-responses-language-model.ts [hunk 2]
```diff
@@ -784,6 +840,14 @@ const responseAnnotationAddedSchema = z.object({
   }),
 });
 
+const responseReasoningSummaryPartAddedSchema = z.object({
+  type: z.literal('response.reasoning_summary_part.added'),
+  item_id: z.string(),
+  output_index: z.number(),
```
Comment (lgrammel) [inline packages/openai/src/responses/openai-responses-language-model.ts]: is the output index needed? i thought it can be reconstructed from the part order

File: packages/openai/src/responses/openai-responses-language-model.ts [hunk 3]
```diff
@@ -133,6 +133,7 @@ export class OpenAIResponsesLanguageModel implements LanguageModelV2 {
       user: openaiOptions?.user,
       instructions: openaiOptions?.instructions,
       service_tier: openaiOptions?.serviceTier,
+      include: openaiOptions?.include,
```
Comment (lgrammel) [inline packages/openai/src/responses/openai-responses-language-model.ts:136]: what is include used for?

File: packages/openai/src/responses/openai-responses-language-model.test.ts [hunk 4]
```diff
@@ -1175,6 +1428,68 @@ describe('OpenAIResponsesLanguageModel', () => {
   });
 
   describe('doStream', () => {
+    const prepareReasoningStreamResponse = (
```
Comment (lgrammel) [inline packages/openai/src/responses/openai-responses-language-model.test.ts]: instead of flags, just set up custom prepare methods or if it's a single test define the test input in the test

prefer less magic / thinking in tests, often it is very helpful to be able to look at the raw input without any indirections / logic

File: packages/openai/src/responses/openai-responses-language-model.ts (hunk 3 above)
Comment (jonaslalin) [reply packages/openai/src/responses/openai-responses-language-model.ts:136]: We need to be able to pass `reasoning.encrypted_content`: Includes an encrypted version of reasoning tokens in reasoning item outputs. This enables reasoning items to be used in multi-turn conversations when using the Responses API statelessly (like when the store parameter is set to false, or when an organization is enrolled in the zero data retention program).
It has also been added to, where I enforce the currently support types that we have also have implementation for in the framework.
```typescript
const openaiResponsesProviderOptionsSchema = z.object({
  metadata: z.any().nullish(),
  parallelToolCalls: z.boolean().nullish(),
  previousResponseId: z.string().nullish(),
  store: z.boolean().nullish(),
  user: z.string().nullish(),
  reasoningEffort: z.string().nullish(),
  strictJsonSchema: z.boolean().nullish(),
  instructions: z.string().nullish(),
  reasoningSummary: z.string().nullish(),
  serviceTier: z.enum(['auto', 'flex']).nullish(),
  include: z.array(z.enum(['reasoning.encrypted_content'])).nullish(),
});
```

File: packages/openai/src/responses/openai-responses-language-model.test.ts (hunk 4 above)
Comment (jonaslalin) [reply packages/openai/src/responses/openai-responses-language-model.test.ts]: Sure, I have now moved it to each individual test instead.

File: packages/openai/src/responses/openai-responses-language-model.ts (hunk 2 above)
Comment (jonaslalin) [reply packages/openai/src/responses/openai-responses-language-model.ts]: I added it to be consistent with the rest of the schemas which also have `output_index` since it is on the received item. Now, everything is consistent and reflects the object we receive from the API call.

File: examples/ai-core/src/generate-text/openai-responses-reasoning-summary-streaming.ts (hunk 1 above)
Comment (jonaslalin) [reply examples/ai-core/src/generate-text/openai-responses-reasoning-summary-streaming.ts]: Yes. I have tested it both with `streamText` and `generateText`. I updated `examples/ai-core/src/generate-text/openai-responses-reasoning-summary.ts` and tested it out too. I have comprehensive tests for both functions to make sure they work as expected.

**Comment by jonaslalin** [comment]
@lgrammel, I missed a couple of complex cases where reasoning is interrupted by tools calls etc. These scenarios have been added to the comprehensive test file. I also removed leaking out summaryIndex and outputIndex, since I think I can reuse the same strategy I did here to reset the reasoning message cache and patch the the streaming case also. At least I will not add it in this PR and do my best to fix it without, so it works for the general case and not only OpenAI provider. Please review again.

**Comment by lgrammel** [comment]
Please rebase against `main` (we have changed v5 into main and v5 will not be updated any more)

**Comment by jonaslalin** [comment]
> Please rebase against `main` (we have changed v5 into main and v5 will not be updated any more)

I have now done a rebase against main @lgrammel .

File: packages/openai/src/responses/convert-to-openai-responses-messages.ts [hunk 5]
```diff
@@ -129,6 +148,55 @@ export function convertToOpenAIResponsesMessages({
...
+                providerOptions: part.providerOptions,
+                schema: openaiResponsesReasoningProviderOptionsSchema,
+              });
+              if (providerOptions === undefined) {
+                throw new InvalidPromptError({
+                  prompt: part,
+                  message:
+                    'Reasoning parts require providerOptions: { openai: { reasoning: { id: "..." } } }',
+                });
+              }
```
Comment (lgrammel) [inline packages/openai/src/responses/convert-to-openai-responses-messages.ts]: this forces reasoning to have ids. is it required, or could we also omit the reasoning instead and add a warning to the list of warnings?

File: packages/openai/src/responses/convert-to-openai-responses-messages.ts [hunk 6]
```diff
@@ -129,6 +148,55 @@ export function convertToOpenAIResponsesMessages({
...
+              const reasoningId = providerOptions.reasoning.id;
+              const existingReasoningMessage =
+                currentReasoningMessages[reasoningId];
```
Comment (lgrammel) [inline packages/openai/src/responses/convert-to-openai-responses-messages.ts]: why would there be multiple reasoning parts (maybe rename from msg) with the same reasoning id?

File: packages/openai/src/responses/convert-to-openai-responses-messages.ts [hunk 7]
```diff
@@ -129,6 +148,55 @@ export function convertToOpenAIResponsesMessages({
...
+                const newReasoningMessage = {
+                  type: 'reasoning' as const,
+                  id: providerOptions.reasoning.id,
+                  encrypted_content: providerOptions.reasoning.encryptedContent,
+                  summary:
+                    part.text.length > 0
+                      ? [{ type: 'summary_text' as const, text: part.text }]
+                      : [],
+                } satisfies OpenAIResponsesReasoning;
+                currentReasoningMessages[reasoningId] = newReasoningMessage;
```
Comment (lgrammel) [inline packages/openai/src/responses/convert-to-openai-responses-messages.ts]: reasoning messages confuses me. it is a reasoning part, right?

File: packages/openai/src/responses/convert-to-openai-responses-messages.ts [hunk 8]
```diff
@@ -99,7 +105,20 @@ export function convertToOpenAIResponsesMessages({
       }
 
       case 'assistant': {
+        // Track reasoning messages within this assistant message by ID.
+        // This allows multiple consecutive reasoning parts with the same ID to be merged into a single reasoning message with combined summaries.
+        let currentReasoningMessages: Record<string, OpenAIResponsesReasoning> =
+          {};
```
Comment (lgrammel) [inline packages/openai/src/responses/convert-to-openai-responses-messages.ts]: why is this needed? when would there be consecutive reasoning parts with the same id? shouldnt that be a single reasoning part to begin with?

Comment (lgrammel) [reply packages/openai/src/responses/convert-to-openai-responses-messages.ts]: i.e. if we generate consecutive reasoning parts with the same id in our response generation, isnt this a bug that should be fixed instead (in a separate pr)?

File: packages/openai/src/responses/convert-to-openai-responses-messages.ts (hunk 5 above)
Comment (lgrammel) [reply packages/openai/src/responses/convert-to-openai-responses-messages.ts]: if i understand correctly, we need encrypted content for reasoning roundtrips. i suggest checking for that before adding, and if it's not available adding a warning and omitting the part.

File: packages/openai/src/responses/openai-responses-language-model.ts [hunk 9]
```diff
@@ -507,8 +525,15 @@ export class OpenAIResponsesLanguageModel implements LanguageModelV2 {
                 });
               } else if (value.item.type === 'reasoning') {
                 controller.enqueue({
-                  type: 'reasoning-start',
+                  type: 'reasoning-start' as const,
```
Comment (lgrammel) [inline packages/openai/src/responses/openai-responses-language-model.ts:528]: why is as const needed now

File: packages/openai/src/responses/openai-responses-language-model.ts [hunk 10]
```diff
@@ -618,9 +650,9 @@ export class OpenAIResponsesLanguageModel implements LanguageModelV2 {
               });
             } else if (isResponseReasoningSummaryTextDeltaChunk(value)) {
               controller.enqueue({
-                type: 'reasoning-delta',
-                delta: value.delta,
+                type: 'reasoning-delta' as const,
                 id: value.item_id,
+                delta: value.delta,
```
Comment (lgrammel) [inline packages/openai/src/responses/openai-responses-language-model.ts:655]: this change does not seem necessary

File: packages/openai/src/responses/openai-responses-language-model.ts [hunk 11]
```diff
@@ -714,6 +746,13 @@ const responseOutputItemAddedSchema = z.object({
     z.object({
       type: z.literal('reasoning'),
       id: z.string(),
+      encrypted_content: z.string().nullish(),
+      summary: z.array(
+        z.object({
+          type: z.literal('summary_text'),
+          text: z.string(),
+        }),
+      ),
```
Comment (lgrammel) [inline packages/openai/src/responses/openai-responses-language-model.ts:755]: does this need to be nulish

File: packages/openai/src/responses/openai-responses-language-model.ts [hunk 12]
```diff
@@ -746,6 +785,13 @@ const responseOutputItemDoneSchema = z.object({
...
```
Comment (lgrammel) [inline packages/openai/src/responses/openai-responses-language-model.ts:794]: does this need to be nullish

File: packages/openai/src/responses/openai-responses-language-model.ts (hunk 2 above)
Comment (lgrammel) [reply packages/openai/src/responses/openai-responses-language-model.ts]: is it used tho? we have a rule to not include any items that are not used.

**Comment by lgrammel** [comment]
left a first set of comments. overall direction is great. can you add a streamText example as well?

File: packages/openai/src/responses/convert-to-openai-responses-messages.ts (hunk 6 above)
Comment (jonaslalin) [reply packages/openai/src/responses/convert-to-openai-responses-messages.ts]: I have attached an authentic response from the OpenAI API below. As you see there can be many reasoning parts that need to be stitched back together again since the `LanguageModelV2ReasoningPart` only allows for a string summary, not an array of summaries. So what we need to do is to process the `LanguageModelV2ReasoningPart`, which is a part, into a Message, which is why I keep track of the messages that has been appended to the messages array. If I think retrieve another `LanguageModelV2ReasoningPart` that really is part of the same OpenAI specific reasoning message, I stitch it together again. The ids are not unique at all but connected to the session, so almost every id that we receive back has the same suffix, but different prefix.

```
"output": [
    {
      "id": "rs_686518ccdb5881a395a979eba2560bc00fbf2733b0e0acc9",
      "type": "reasoning",
      "encrypted_content": "gAAAAABoZRjn4Ej2m_Pjl4aaPYBVWnwzxGNRbGfYK2YjqyAKpA-rFG7se1oe699VdivCrwlbnbPwi0j1c_Ds35QvNLepFj19j6Ui0gh3ri1cUuSbDePkKOG46Pf4xz3hbfWPPm9_4qX6ZWIg1RejWWRQyvAekd8N-oBSSkT5AWzpoYgW-HOqMoREvWRE4KJUGV_b4fBSEuIDvBOC5KJ9tECZTbc4Pqpp-NXrz2uGeLLhGoTbuvnMq_jX8FHMvgMWrLp4lur-mjF31h2PNN6AOM3ghTVu4FTCkdI83JQLFzmD8wCFE4U3iGr-nIIBI2UsAH4h24PHqOM0eT72BBCIdx9EpL85OQD60e7wXsj5RnWzJQBWuq6C5BSWMgR4V38sfN-z3pyDyyb1zsROD235phcE1_uEeSe0dSIFqFGNN4VSgY_hJtbYR8b_WAY3WXpxGUB6qEyhhLqHJx17_LyxhHlm6cxH3cyZmBsb2O3LV29rJt7rpNa_UlB15qLsP1WuAgSZZrdukCY62cslfzC0xRHFN2DueMErXEFdsD6hDcLwkcH-kmd_8Zeni-xlUCkXhS_P07z1tjwIO0VHCyEVQfpZ-U42HQUpD17sAubktgiOyB40owyCecKl_-gqvw9FN-5CCzsyD0Ot7bkKntaDreNMSEHl5c9z1M9jqeSd2Ql0tdcw5R56CuzH_CtrdywdWfILHUniDC_7Qri3TFJbsvRHfwjrO689gy32qBcstxCbv1nDp3GqV8P4CXLcHYXgcg1ghEnqGo_76DJF5H1LjoNtR91zV_ZIYCnf3JDuVfPffD5JVObIUdsy6KEe-P6bVu_prjQzuWdrD5GDkAKZ8EfsSOxlODiDDrb6_jMDic8VRcWpTbAZsQWrXFUKzasc9Urge3rRXZqinFp6mag4D5U-hTkPdF2DlbugGEEBVVCJ-6CsE19-eWECaes7iu1J9cBx43QvtbK-K9DU-msL0uDGxrxlGFMOWkZXTsoN3ZL7xP24nwIuaEnMN9EieKLk9hCR3hMw8wjRcExh-GWU_ns_c7BjJo-xV8xbz7FGDTus6oenAUQI5mbJqM43Ye-KfLXfdeub5zH0blV6-sa2qSB4R2cObLNQ8DghTfdB_YNVLSz3Y2BHE9XMcF2GW0uQ7xDJp83k4H_KJ0hw5-YOj2xnphORO0JlcGDEX8IxmtEFXte3jfIUy5ISVnTDcR8BFGT8C376_LsxBR3vdPwQBB505ksP5kzj-DMq3qAI_bae-IzKuYX9dS09M1tdzfrXitJLF_r7-ONKOvnj8xn8x6bw6kCHLop_BaPqAB4693rGq-3tTz7pTOYuy-9JSiKcPYYatdGosjCqQjoTBURPJrLcLq_mrr_JAdUZcvR7YKjFW-K6z1WSCF4Cq5Rb9rF3mMYLIXIQvGSxIfNKf-fWvLyDyIUKSukQqtLeBU8Q4JhrCtpWuspMsscOABDCPxzU4sQCOVbZAr1vEmwnEVwVUwiwGkG4Tcwvi9R8TWPyIteQtQ3qpB0ON-CGcTf-6dcWaMH6vSjKCZjJwVecxAXxIJQnpam3U6IplbpzUVotmWtiJPlLxiw2oN56E5Xt187QbIjTORtIum-ogUFj_VrCzTXO-Imy8qIKbo05ttlb3qXH3m5Nbilrt0UOA0ysfES0-kAutDAqxDZZVLZK01DYTv9vL2y2x50hrpgbV8_5I5Ny0hMeBFOjRNjIQw_03mxKGCTMzV_6y3B9DoCev6KhnVmWdZ1mYTyOhL9CBH9vID8tTZ6zFdCfc7WLfeseX3KTQzqyK0MlN0cvkSd91X5FVVfLjc8ZnphcHOq2qOVL4ZrqCPGc1ySYmRDi0FJLmMo1qYm8UPQZXHh9u39vgr-NR0Z6kRAGfx_PnnXA4GQUJ6fU4H2aQZDBvphi7QIadFoTj6v5hBlZAfwuvEIBedAxtu1Rnpb6n9RRjUR4QDkNH9qWBEtQgk6Tt2-SLlwo09xSbgk9_ycgWuuLR3IEGjN13_6cDPy-ZZQ0iQZGLnSnpfM0jHXk8NQDo1OZidhVKaZ9mW_cRPaXB0f-0EogkAz4onNQIrKkZRj32IJwD5k6IxW3mP1XXE7os07oXJimXtBy-SVZfB5P4sW2VMcQgPAZrxNT4hgkn71PNCQzkuVsZdBz2VJ83jzLsS0vR020BId9azZKdo9VFOiI0wg_hZwyrszha9lrPgW0f3g3qXaru8ZG01unEN78xKeOSVRv8ovhkv7Xo5_RjRw6NzWAVBo4-d3WzavOYW88HlISMSj_XNvifdHRtSyRCXBzmDYZr60VM3HkeMGmhcQxaFgkl7OUo6a_-DQCvnRmMpPbzyAmXdFuccZluog3G0bzQOoBp4Cc9hvnD24Aqyuqxqg88T-qKi0MCRgFjAvicgAHXkPy-4R6NjNnRTw3_jmkgg5_BCjl4sCEzqbWzjGZ3fZom72RBcDzmHMh9AL4Twv9zhEsjzM7BLpy7i8IdK471TYFL-QgAYvgAzcIn88EdbwsTNVpHpEis-abGOFW1RN1VTWY4d-zfHb292iBrQ3_hwhAZ-_APcCcgZZcfrUla0xoqeHMMquGpPGwQKR7-OE-bFoAH0GoAgbpYOPIn9csYc9jCUI5fp08rHZbX_F0KZ3Ydz5D9fuqzcWHf1dsNd7Je_GMigV_5cTvQFsVlHmLIxk0HcFUs2DIVF2ReXTWEEyG2UmGuGamjMpXvUc87bA8w8pjbCPxIom9MuSGrdgE9B7Egcqet9r5ed2qGYoei-x8xSoIQWlt8BQASPmSYoKONrCHR-evuPLTH3siUeIOgDmA_khPDSj_AkSFzWgzjV5OGwrPna5L7TOa2vGc7vH7kmPIHrvverhf38qVJwbTFqNuVo_0Cc369fnLkINQkGrEqrL8EG5v-kZcWT53Hs_fWXhIhCsJxj7XWvcBuOnu5jF7ze-qCUMFtLWRzq_HHmle5gvEFgeQFqFJcXEOlBH-vM6qS0nsvzNtkNiTDx0j4yxWOy-PH2VymFKmLhYJ86I4r-9ul-BN1jXXHYQAuvSO2bSt4wc8QBP4z9eMHxbWFyyc_NnV2GOI6AEfF6bZMYpElVBn6kSkhijQuk8t9ZPeWL_DqVLJgQVH0uqeh5k54rBOkdhCFkhobWmT-Gqwo6sC7coOlt7JHepsHUbQ10cofGgsNUj4k1K3tpe00DOdp_XeGjI4FqKwvqW1FY-5ybEXlrPI350zh5hKFZKV83foDeTKYAiQExcvTbaQ_A7KpftxDxrGJvVAvEltpm25ymp0np_Owaa6MjrUmwq6mL93WZ1hTNUEWPurJxgeuWa8qvgcF0XXJ7WODxSNI-4qk3OIjNZiy5AfAitqjboEXlQL0mXeafBU_ZJVJhWpJR0a9izZlBycnaqPZzwqmIbVzglA0HeBzdgOq7rtDl9W536_ndanGvk98aA-3R5PirBl6Zgw5LGz9P-Gya6YCUFLLlZ2YPq3Nwv8_ficnZ0VhM4o1E3vHRPVcma0qHe1GhtS7MBKB35Vc9TwYDdJ3v8Q_nLndiJ8_vVbwoAZj-mXZjLR96Vf72lMNQvNq7okn2pn9lBpjmOdxmWShiBn0hchhKTjtE6k1s6SKhMzgqSdz-0K7kcmABGqW6Ap9a5sr8fBIddI62_1-O0Wq5tMWom5dripwzIYhA7128qyGZBoK-QQYQkPzWPoKsST8rhpC4ElGyZs4ougZrxsbWRH1zfFRfLz6on_NQrinS0Qby0Q96AhegPvEiyWQaLYv0WZ7lP0WYwvLJ25Umo7IXoUZH9xWUaIDAQgtPSYJymL-jym5_iA99s3Qe9Z_IxMN1THXa5tsk3hSIUkXicAz9bQ1YVOyohCQ5oIp-P38Ma1HJSCERp9NTXRv9yqKwXaubnx8W_cOF3JrlxgZ2F2flJUIPdr2Yz0xO4rRMaEOuZRlIN5LCi7RS5DbIKnXRIpWiNHYQBrNqnU_kabbkO7cKump_sAxm5457GF8hweqqiyL7Pupy0s0ab6DIit0sHHLFukXN6WCByLjvBkiWFyacNxzzMEg3pIiw5ydrx8XFcYCCaNxqQQapIDUSgRu2gjTEnFZr4o-HZqSaFeWx1kt2-hYufD2Au8G1hDOWjF2uyqoLe8gq9N8B5Pu_FyIBV6dnFEhqW2AsZptRZD0Uda-PymGSZbEkFn2hTlOHwnYTK6_GtMSz1P5CXq3_ovXrYEzb-4U02dvMpjoqGdVMvUWcsZ4VlkAxdMYO2cCqn5mcRUqP2YpW0VRVAGV2AXKJh-qjaw2tHgC_1FBLvoxnjtOAYVWnzXMHohusnSanub9p1JvH_Sy7LEW48KKZq4fafqhjxBwWL4MIHvcZfmPzTcHkZkd0tdiqVzBVGTE8S-paJ11oSyZn-zX2Of8s3kiXPvMRMIGO_fuIs_kK_dxZfcjXC8wptRo-ZlygeYR_jjAV-j9oBSOGEshQghBygER5g6M2f5ZE84iRuAIRrcGqwSulTkowcFMVVIaEDRYYjSrohRmvF0x8CMijiC8bcKcaren5lNVV7_IDroi-gfnm0fHUxGcK5pwyi2Db-IVhbu7NdsX1HBra82J2tIalYdhvZa5SGE6hI2ToqKHjSc1AVBpBMwKi71BUzFu-f4TnIMlMsZQHebpBul80lvYkvnwG7zBbS0xQvIoF3MxstEGszTTn5HuPEy5fmOqZjja1uwNCgZUWs-rnlkSO-_eSlFwrzZPaV3k8YEVlHBk37PCwKj8SIvEuCffNX-lTG9EY6OG2ul3NaZoS5VzlET0qx62okiCE0vgYpsg-glCXAjDSS6eTcPPR2bRY0oK12r7UX7YLgutldLUaZJESbnuA3tp5ERZWpfKgWYPzhu3cSztv1cnGdKzpds8bpIRUVmLBP1Pn4etMiZJ_5dvYKEwKm_NAHttU41MGOkVhCaR0laUp4If4rCNVnX0___JI2elrbyYMw5zWENB7jvCeVL7A5Oe4uIvsGP6gXzW3Lq4RvRr7XpaxLFyp7kQ6BvsN9OpY71VgJgtIGT_O7fx7Upm2kb3L4sgN_L8w9gPV_LFpSaKfIjp9cQo_BQkuwpge7QhEfUPK9-uesfnUb6cUqaRgK0L3XspJQ1_Kag357SOTtFLLIOTVdyLBn6EsnhQS6eiV6uE2AWQaXyuQ==",
      "summary": [
        {
          "type": "summary_text",
          "text": "**Planning a trip to Paris**\n\nThe user wants help planning a trip to Paris, which I find exciting! I need to clarify some details first, like trip length, budget, and interests. I could suggest a sample itinerary covering attractions, accommodation, and local tips. It's important to ask about their preferences, whether they're traveling solo, as a couple, or with family. After gathering this information, I could provide a detailed 5-day itinerary with highlights, transport options, and budgeting advice. Let\u2019s get started!"
        },
        {
          "type": "summary_text",
          "text": "**Requesting trip details**\n\nTo assist the user with their trip to Paris, I need to ask some clarifying questions. I\u2019ll inquire about trip length, travel dates, budget, interests, and preferred style. Once I have that information, I can provide a detailed itinerary. I might suggest a sample 5-day plan that covers main attractions and costs, while leaving room for customization based on the user\u2019s preferences. This way, I can ensure the plan is tailored to fit their travel style and interests!"
        },
        {
          "type": "summary_text",
          "text": "**Providing travel tips**\n\nI want to offer some helpful travel tips for the user\u2019s trip, like suggesting museum passes to avoid long queues and sharing dining etiquette essentials. It's important to consider those small details that can enhance their experience! At the end, I\u2019ll ask for any modifications, such as trip length, budget, and specific interests, so I can tailor the recommendations even more effectively. Let\u2019s make sure everything is clear and concise as I finalize my suggestions!"
        }
      ]
    },
```

File: packages/openai/src/responses/convert-to-openai-responses-messages.ts (hunk 7 above)
Comment (jonaslalin) [reply packages/openai/src/responses/convert-to-openai-responses-messages.ts]: I use the word `Part` when it comes from the incoming Prompt, but we have `const messages: OpenAIResponsesPrompt = [];`, so that is why i used message when I attach something to that array.

File: packages/openai/src/responses/convert-to-openai-responses-messages.ts (hunk 8 above)
Comment (jonaslalin) [reply packages/openai/src/responses/convert-to-openai-responses-messages.ts]: See example above for real response using the API. We dont generate ids at all, OpenAI returns them and they use the same Id for a lot of things.

File: packages/openai/src/responses/openai-responses-language-model.ts (hunk 10 above)
Comment (jonaslalin) [reply packages/openai/src/responses/openai-responses-language-model.ts:655]: All other types related to reasoning have the order type, id, delta. So i use that too to be consistent. I think that looks better than mixing.

File: packages/openai/src/responses/openai-responses-language-model.ts (hunk 11 above)
Comment (jonaslalin) [reply packages/openai/src/responses/openai-responses-language-model.ts:755]: No, it will never be. It is always an array, that can be empty.
image

File: packages/openai/src/responses/openai-responses-language-model.ts (hunk 12 above)
Comment (jonaslalin) [reply packages/openai/src/responses/openai-responses-language-model.ts:794]: No, see answer above.

File: packages/openai/src/responses/openai-responses-language-model.ts (hunk 9 above)
Comment (jonaslalin) [reply packages/openai/src/responses/openai-responses-language-model.ts:528]: I always use const since it plays well with discriminated unions, type hinting etc. I would say it is recommended to always do that when emitting something that is a literal and not a string.
```
// Without 'as const' - type is inferred as string
const event1 = { type: 'reasoning-start' }; // type: string

// With 'as const' - type is inferred as the exact literal
const event2 = { type: 'reasoning-start' as const }; // type: 'reasoning-start'
```

File: packages/openai/src/responses/convert-to-openai-responses-messages.ts (hunk 5 above)
Comment (jonaslalin) [reply packages/openai/src/responses/convert-to-openai-responses-messages.ts]: Yes, it is required. I have followed the spec. I can change to omit the part and add a warning instead.
image

Comment (lgrammel) [reply packages/openai/src/responses/convert-to-openai-responses-messages.ts]: there are cases where users switch between providers in a chat, or construct their own payloads. the assumption that this is available does not hold. you cannot assume that the reasoning part was produced by openai.

Comment (jonaslalin) [reply packages/openai/src/responses/convert-to-openai-responses-messages.ts]: To me it makes no sense that LanguageModelV2 requires ids on doStream, but not on doGenerate? If I reconstruct my state from the stream, I have ids, but if I have called doGenerate, I dont

Comment (jonaslalin) [reply packages/openai/src/responses/convert-to-openai-responses-messages.ts]: So why does everything have an id in the doStream case?

Comment (jonaslalin) [reply packages/openai/src/responses/convert-to-openai-responses-messages.ts]: I could make it an optional property and generate the Id for cases like this, but I don't know if it will improve the reasoning of the model, because the model is only using encrypted reasoning and doesnt really care about the plain text text summaries. I guess that is true for many providers, which dont share the actual reasoning, but just give you a summary, and an Id to pass back so they can fetch the internal reasoning on their side.

**Comment by jonaslalin** [comment]
> left a first set of comments. overall direction is great. can you add a streamText example as well?

@lgrammel I have now answered your feedback. I have changed so that it is possible to pass reasoning items to OpenAI without Ids or any other providerOptions (even though it doesnt make sense to do so...)
I have a small streamText example in `openai-responses-reasoning-summary-stream.ts`, did you miss that?
I will add a comprehensive one, once I fixed the stream code in the core provider package.

I really recommend reading the tests I have added to better understand how it works and what edge cases I am also solving.

File: packages/openai/src/responses/openai-responses-language-model.ts (hunk 2 above)
Comment (jonaslalin) [reply packages/openai/src/responses/openai-responses-language-model.ts]: it is not used for any of the object we have, there, either i add it, or i remove it from all the other openai reasoning parts as well?

Comment (jonaslalin) [reply packages/openai/src/responses/openai-responses-language-model.ts]: I have removed and cleaned up the things we dont use there now

File: packages/openai/src/responses/convert-to-openai-responses-messages.ts (hunk 5 above)
Comment (jonaslalin) [reply packages/openai/src/responses/convert-to-openai-responses-messages.ts]: Now done

**Comment by lgrammel** [comment]
I am continuing the work in https://github.com/vercel/ai/pull/7014 no need for further changes here.

**Comment by jonaslalin** [comment]
> I am continuing the work in #7014 no need for further changes here.

Sure, let me know if you need any clarifications for how I was reasoning :)
